sqlx-core/src/postgres/listen.rs

```diff
@@ -0,0 +1,297 @@
+///
+/// Notifications will always correspond to the channel(s) specified when this object was created.
+///
+/// This listener is bound to the lifetime of its underlying connection. If the connection ever
+/// dies, this listener will terminate and will no longer yield any notifications.
+pub struct PgListener<C> {
+    needs_to_send_listen_cmd: bool,
+    connection: C,
+    channels: Vec<String>,
+}
+
+impl<C> PgListener<C> {
+    /// Construct a new instance.
+    pub(self) fn new(connection: C, channels: &[&str]) -> Self {
+        let channels = channels.iter().map(|chan| String::from(*chan)).collect();
+        Self {
+            needs_to_send_listen_cmd: true,
+            connection,
+            channels,
+        }
+    }
+}
+
+impl<C> PgListener<C>
+where
+    C: Connection,
+    C: DerefMut<Target = PgConnection>,
+{
+    /// Receives the next notification available from any of the subscribed channels.
+    pub async fn recv(&mut self) -> Result<Option<PgNotification>> {
+        loop {
+            // Ensure the current connection has properly registered all listener channels.
+            if self.needs_to_send_listen_cmd {
+                if let Err(err) = send_listen_query(&mut self.connection, &self.channels).await {
+                    // If we've encountered an error here, test the connection. If the connection
+                    // is good, we return the error. Else, we return `None` as the connection is dead.
+                    if let Err(_) = self.connection.ping().await {
+                        return Ok(None);
+                    }
+                    return Err(err);
+                }
+                self.needs_to_send_listen_cmd = false;
+            }
+            // Await a notification from the DB.
+            match self.connection.receive().await? {
+                // We've received an async notification, return it.
+                Some(Message::NotificationResponse(notification)) => {
+                    return Ok(Some(notification.into()))
+                }
+                // Protocol error, return the error.
+                Some(msg) => {
+                    return Err(protocol_err!(
+                        "unexpected message received from database {:?}",
+                        msg
+                    )
+                    .into())
+                }
+                // The connection is dead, return None.
+                None => return Ok(None),
+            }
+        }
+    }
+
+    /// Consume this listener, returning a `Stream` of notifications.
+    pub fn into_stream(mut self) -> impl Stream<Item = Result<Option<PgNotification>>> {
+        stream! {
+            loop {
+                yield self.recv().await
+            }
+        }
+    }
+}
+
+impl<C> PgListener<C>
+where
+    C: Connection,
+{
+    /// Close this listener stream and its underlying connection.
+    pub async fn close(self) -> BoxFuture<'static, Result<()>> {
+        self.connection.close()
+    }
+}
```

Comment:
Not sure what you all think about this, but we could wrap the connection in an `Option` so that we can take the connection out and spawn a task to call this `close` method from inside of a `Drop` impl. That way user's wouldn't have to worry about forgetting to call `close` on their own.

With the `PgPoolListener` that is easy (already implemented below), as it wraps the current connection in an `Option` for internally required reasons.

Thoughts?